chat UI design improvements ("Cody Tomorrow")

sqs opened the pull request:
Implements some of the chat UI design improvements from [Cody Tomorrow (Figma)](https://www.figma.com/file/xZ7JeITOwyq2eqV4H0ePVN/Cody-Tomorrow?type=design&node-id=2-3117&mode=design&t=aMcNx4NS46ORwalF-0):

- Each "cell" in the transcript (called this because it's starting to resemble a notebook instead of just a chat session) now has an avatar representing who it's from (human, context fetching from Sourcegraph, assistant).
- Context is shown in its own "message" in the transcript.

This is only a subset of that new design, but it's a subset that looks awesome and makes sense to ship incrementally.

## Test plan

Existing e2e tests and CI. Also try all of the edge cases in chat: aborting messages, editing messages, code blocks, error messages, etc.

---

![image](https://github.com/sourcegraph/cody/assets/1976/3a464b55-d7dd-4834-b4e0-7a9bbcf1feef)

abeatrix reviewed (commented):
Looks great: 

![image](https://github.com/sourcegraph/cody/assets/68532117/bfb7d87c-08b9-406f-ad20-57c5d9d2fcd5)


Noticed some minor styling issues from the build:

The edit button's height/position on some questions:

![image](https://github.com/sourcegraph/cody/assets/68532117/97ce6f81-7b86-401a-86fc-0b24985a9ac8)

The context icons and files links are underlined:

![image](https://github.com/sourcegraph/cody/assets/68532117/03424ff9-bb36-4c0a-90ca-8cc740ec1961)

![image](https://github.com/sourcegraph/cody/assets/68532117/83133441-199a-4daf-9c31-df009f30c908)

Tested on s2, demo, and dotcom and everything else looks great to me. Tried to log in with users with and without avatar and the chats still work as expected ✅ 

Will defer to @toolmantim for the final check!

chillatom commented:
This is awesome! 

One minor nit is that when the user submits a message it first shows a cell with the LLM logo and no animation, then the Sourcegraph logo cell appears with a loading animation as context is fetched.  The delay is minor, only 1-2 seconds, but there is a period where it is unclear if something is working. 

https://www.loom.com/share/563cf175e84041659d6b72e537f35c22?sid=baeeb77d-a964-4051-8d90-8f1d0b0e957d

sqs commented:
Thanks @abeatrix.

Fixed the context paths styling. (This is on my color theme so it looks a bit different, but the problems you mentioned are fixed.)

<img width="456" alt="image" src="https://github.com/sourcegraph/cody/assets/1976/a6c49523-d643-4fb0-a72b-e4651d15ebce">

And fixed the tall edit button issue:
<img width="430" alt="image" src="https://github.com/sourcegraph/cody/assets/1976/5f0def9f-7d94-4067-9ade-3e2d4092691d">

sqs commented:
@chillatom Cool, thanks. I see a lot of jank in that animation. Will work on it!

sqs commented:
@chillatom OK, fixed. It now is smooth. It shows both as loading, which might be OK and might not. I'll wait for @toolmantim to confirm here.

https://github.com/sourcegraph/cody/assets/1976/b87b173a-5dcd-4dbe-980b-f7bdb9775137

sqs commented:
Added full storybook and React component unit tests for all the different permutations of the transcript (context loading, assistant message loading, no context, etc.).

toolmantim reviewed (commented):
Very cool!

There's a few small problems in the original Figma which became apparent when looking at it live:

* Avatar is 24x24 but Wildcard and LLM icons are 20x20 — let's make them all 20x20 (or if they're smaller like Anthropic logo, centre them in a 20x20 container)
* Context chevron should be 16px not 19px with 4px margin to the right. Font weight of "Context" should be 500.
* Let's not do the total count badge for Context I think, and instead bring back a count of files and lines: https://www.figma.com/file/xZ7JeITOwyq2eqV4H0ePVN/Cody-Tomorrow?type=design&node-id=172-4678&mode=design&t=hUSNq16YFz3AmwNI-4
* Let's bring back the "source" icons + tooltips from #3528 showing where context items came from ![image](https://github.com/sourcegraph/cody/assets/153/aa568fee-ad62-47b0-a511-0c10212e32b1)
* Feedback buttons: let's just keep them visible at all times, or if you really want we can fade them to 25% until hover
* Edit button: let's keep that visible at all times too (especially while we have follow-up questions context length problems)
* Unrelated to this branch, but I could I sneak in a fix to remove the extra padding on the message section when there's no buttons:  ![CleanShot 2024-04-02 at 22 31 42@2x](https://github.com/sourcegraph/cody/assets/153/e0a24f42-fe4e-485f-b7ab-326d388d22bd)
* Is it possible to unify the style/display of @ objects between the human item and the context list, because now they're quite different: ![CleanShot 2024-04-02 at 22 33 47@2x](https://github.com/sourcegraph/cody/assets/153/e293057e-708c-479c-a67d-cdfe8aefef3f)

abeatrix reviewed (commented):
Looks great! 

I run into a bug where clicking on one of the old chats, the blinking icon would show up. It doesn't happen to all chat though:

https://github.com/sourcegraph/cody/assets/68532117/005c51ad-3414-4810-824c-dc68f2dc33dd

Also notice the following changes but not sure if they are intentional:

Local context is highlighted with a different padding:
![image](https://github.com/sourcegraph/cody/assets/68532117/9eff3ed3-d56e-4386-be5b-9ba90b54edec)

The context icons are not updated based on source, expecting: 
![image](https://github.com/sourcegraph/cody/assets/68532117/339ee92d-73df-44cb-aa2f-4cd023338975)

Excluded @-files are not displayed in UI
![image](https://github.com/sourcegraph/cody/assets/68532117/8cba1b99-8daf-4f44-b012-19e2944c778d)

abeatrix reviewed (approved):
Comments inline, mostly for the styling on FileLink. Didn't notice any regression. Ran into the bug where opening a chat would show that context is loading, but it was only that one particular chat so probably user-error :D 
I think this is a great start and we can refine the Context List more in follow-ups?

sqs commented:
> Also notice the following changes but not sure if they are intentional:

Fixed: 
![image](https://github.com/sourcegraph/cody/assets/1976/581484a1-bc34-4fe8-88c2-6709b605e7b3)

sqs commented:
@abeatrix: What do you mean by this?

> Excluded @-files are not displayed in UI ![image](https://private-user-images.githubusercontent.com/68532117/318924681-8cba1b99-8daf-4f44-b012-19e2944c778d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIxMTc0MzAsIm5iZiI6MTcxMjExNzEzMCwicGF0aCI6Ii82ODUzMjExNy8zMTg5MjQ2ODEtOGNiYTFiOTktOGRhZi00ZjQ0LWIwMTItMTllMjk0NGM3NzhkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAzVDA0MDUzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRjZWU2YzZkMTQyNTdkMmJjMWEyZDhkMzc5YmMyYmFjZGUwNzNkYTQ2NzQxZTE1ZjViZGQ5OTNlNWZlODRlZTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iIkej6ct14njbU_jSUUizDCf5tZB8rUyJhQHtqaKQ2g)

Are you talking about `.cody/ignore`? When I tried that on the nightly build, I see the same behavior (`cody.internal.unstable` is true and this file is .cody/ignore'd):

![image](https://github.com/sourcegraph/cody/assets/1976/dcd5c1ee-7005-4903-8a0d-b182d500b13f)

I agree a better behavior would be to display /why/ the file was ignored, but that file (as best I can tell) is entirely excluded from the `ChatMessage.contextItems` array, so we don't actually even know that it's in there in the context UI. I think we will improve this behavior when we work on Cody Ignore for Enterprise, but for now, I believe the behavior is the same.

Let me know if I missed something, though!

abeatrix commented:
> @abeatrix: What do you mean by this?

Hey sorry, I forgot to update my old comment. That was user error on my part. I can actually see the excluded @-mention file showing up correctly: https://github.com/sourcegraph/cody/pull/3639#discussion_r1548726871

sqs commented on vscode/webviews/Components/FileLink.module.css:
```diff
@@ -9,12 +9,20 @@
+    display: flex;
+    align-items: center;
+    gap: 0.125rem;
+    overflow: hidden;
```
ContextCell applies this background color. I thought there might be other places where we just want a FileLink rendered outside of a ContextCell and would not want that bg color. There aren't now, but there have been in the past.

sqs commented on vscode/webviews/Components/FileLink.tsx:
```diff
@@ -69,9 +67,8 @@ export const FileLink: React.FunctionComponent<FileLinkProps> = ({
-            <i className={`codicon codicon-${icon}`} title={getFileSourceIconTitle(source)} />
             {/* biome-ignore lint/a11y/useValidAnchor: The onClick handler is only used for logging */}
             <a
                 className={classNames(styles.linkButton, isTooLarge && styles.excluded)}
```
Why remove? I thought we'd want to show a ~~strikethrough~~ on a file that is excluded?

sqs replied:
Ahh, in conjunction with your other change below. Makes sense!

sqs replied:
Ohh, nevermind, this is removing the strikethrough from the icon. Which makes more sense. lol gotcha now for real

sqs commented on vscode/webviews/Components/FileLink.tsx:
```diff
@@ -80,22 +77,13 @@ export const FileLink: React.FunctionComponent<FileLinkProps> = ({
                 onClick={logFileLinkClicked}
             >
-                {pathWithRange}
+                <i className="codicon codicon-file" title={getFileSourceIconTitle(source)} />
```
done

sqs commented on vscode/webviews/Components/FileLink.tsx:
```diff
@@ -80,22 +77,13 @@ export const FileLink: React.FunctionComponent<FileLinkProps> = ({
             >
-                {pathWithRange}
+                <i className="codicon codicon-file" title={getFileSourceIconTitle(source)} />
+                <div className={styles.path}>{pathWithRange}</div>
```
Having the `@` or file icon be white not blue makes sense for excluded files, but not for included files. Most files are included, and IMO it is better to have the `@`/file icon be a consistent color rather than sometimes white and sometimes blue. So I think on balance the `@`/file icon should be blue (link color).

sqs commented on vscode/webviews/Components/FileLink.module.css:
```diff
@@ -9,12 +9,20 @@
     font-weight: normal;
-    display: inline-block;
+    display: flex;
+    align-items: center;
```
I can see very small differences, but the "Before" looks better to me. What do you like about the "After"?

abeatrix replied:
I like that it's centered, but that's very objective, plus it doesn't affect ui or ux, so you can ignore this XD

sqs commented:
@abeatrix :

> I run into a bug where clicking on one of the old chats, the blinking icon would show up. It doesn't happen to all chat though

The chat you picked is an interesting case. It is a chat with a human message but no assistant reply. Curious, how does it render in the nightly or stable extension version?

I tried to replicate the ways we infer the "state" of a transcript from the object itself, but it's tricky and there is probably a bug. (BTW check out `Transcript.test.tsx` for my test cases.) I would guess that the transcript in your case has a `ChatMessage` with `contextItems: undefined`, which we interpret as a loading case. If this branch breaks it, it's definitely a bug. But I can't seem to repro it myself:

![image](https://github.com/sourcegraph/cody/assets/1976/ddaa7a5b-f3a8-477b-9603-f7d2a5ac0604)


 Would you be able to export that chat and send me the JSON of just that transcript that you get from `Export Chats as JSON`?

sqs commented:
@toolmantim I made the avatar 24px, the Sourcegraph icon 20px, and the LLM icons 18px because that looked right to me and that's what your Figma actually has.

Here is where they are all 20px. IMO the avatar looks too small, and the LLM icons look too big:

![image](https://github.com/sourcegraph/cody/assets/1976/1eb4ca33-126b-4f6d-adde-91a0eb453170)

sqs commented:
@toolmantim:

> Is it possible to unify the style/display of @ objects between the human item and the context list, because now they're quite different:

How far do you want to go in unifying them? I can do this, which uses the same foreground and background color:

![image](https://github.com/sourcegraph/cody/assets/1976/1823948a-c1fc-4037-b4f4-0c92dd350b4b)

![image](https://github.com/sourcegraph/cody/assets/1976/64f4860b-f305-47df-9533-a73f9e2ff9c2)

The padding will be different, though, because we display the editor chip as text. And the icons aren't there. If we add padding, it starts to look weird and non-text-y. (We might be able to make that feel good, but that would be more work.) We can add icons, but I'd prefer to leave that out of scope of this PR (and the only icon that makes sense right now is `@` anyway since the file icon is for auto-included not at-mentioned files).

sqs commented:
@toolmantim All other design comments addressed in this branch, except the one about the context file icons.

1. We don't have the nicer `<>` file icon you have, only this one from codicons that looks not great: 
![image](https://github.com/sourcegraph/cody/assets/1976/e38137a3-efd2-4d7e-ab7e-dedbd00d5fe4). We can get the nicer one in, but I'd prefer to leave that out of scope of this PR.
2. I assume you mean removing the dual icons (except in the case of large files) as in your `More minimal` Figma frame.

sqs commented:
I will wait until AFTER the release is cut to merge this to `main`.

sqs commented:
@toolmantim Couple minor questions from me to you blocking this, otherwise it's ready to merge.

toolmantim commented:
👍🏼 Yeah let’s get it into nightlies!
